### DISCUSSION TITLE
One to Many and One to One between same table

### DISCUSSION THREAD

**Discussion by OddBlueDog** [sqlalchemy/sqlalchemy#6870] [Usage Questions] [answered]
What is the correct way to have a one to many and one to one between two tables? I have created a simplified version below of the the issue I'm having, where a user can have many books and one favourite book.

When I do the following and create a migration using alembic I get an SAWarning saying _"SAWarning: Cannot correctly sort tables; there are unresolvable cycles between tables "user, book", which is usually caused by mutually dependent foreign key constraints.  Foreign key constraints involving these tables will not be considered; this warning may raise an error in a future release."

```
class User(Base):
    __tablename__ = "user"
     id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    favourite_book_id = Column(Integer, ForeignKey("book.id")

    books= relationship("Book", back_populates="user")
    my_favourite_book = relationship("Book", foreign_keys=[favourite_book_id])
```

```
class Book(Base):
     __tablename__ = "book"
     id = Column(Integer, primary_key=True, index=True)
     owned_by_user_id = Column(Integer, ForeignKey("user.id")
     
    user = relationship("User", back_populates="books")
```

Edit: fix various syntax issues

**Comment by OddBlueDog** [comment]
I have added use_alter to ```favourite_book_id = Column(Integer, ForeignKey("book.id",  use_alter=True)``` and the warning is gone now. I don't fully understand why use_alter resolves it if anyone has any further insight.

I also added a primaryjoin to ```books= relationship("Book", back_populates="user", primaryjoin="User.id==book.owned_by_user_id")``` and also an explicit foreign key to ```user = relationship("User", back_populates="books", foreign_keys=[owned_by_user_id])```

**Comment by zzzeek** [comment, marked as answer]
the ORM docs that I think you followed here should probably include a link out to the Core section on what's happening with the foreign keys, which is here: https://docs.sqlalchemy.org/en/14/core/constraints.html#creating-dropping-foreign-key-constraints-via-alter

the short answer is relational databases have two ways to create a column with a FOREIGN KEY constraint on it, there's "inline" and there's "alter", the latter of which is after the fact.  when you have mutually dependent foreign keys, at least one of them has to use the "alter" style.  you can't use the "alter" style in SQLAlchemy without giving the FK a name.

if you are using alembic, you will want to use the [op.create_foreign_key](https://alembic.sqlalchemy.org/en/latest/ops.html?highlight=create_foreign_key#alembic.operations.Operations.create_foreign_key) directive to actually create these constraints, after the tables have been created.   this would be more of an alembic question, alembic currently doesn't automatically figure out this condition when you do autogenerate so you have to set it up manually.
